storagenode/storagenodedb/bandwidthdb.go

```diff
@@ -197,6 +182,7 @@ func (db *bandwidthdb) Rollup(ctx context.Context) (err error) {
 		SELECT datetime(strftime('%Y-%m-%dT%H:00:00', created_at)) created_hr, satellite_id, action, SUM(amount)
```

Comment:
1. The date-range probably should come as an argument instead of internally computed.
2. The coalescing interval as well.
3. This code should assume that there can be bandwidths in bandwidth_usage older than 1 day, and they should be properly handled.

Reply:
1. This select groups by hour, satellite, action.  So it's necessary to have it done in the select so we can handle multiple hours of followups at one time
2. This could be passed in, but it would require 2 trips to the DB
3. Added a test to show the rollup handles data older than 1 day

Reply:
Based on the code I'm not able to understand that it handles it correctly. Let's say I put data older than 1 year into bandwidth_usage table. The rollup won't touch that, since the `created_at` never matches the query. Then the `DELETE FROM` part will delete it.

So this suggest to me the tests have a bug somewhere.

Reply:
First time the rollup occurs it rolls up all data up until the last hour.  Correct, all subsequent rollups will ignore data that was added not with the time range.  But that would suggest some process added more bw usage over an hour after the upload occurred.  Is that possible? BW usage is added after each piece is uploaded.

Reply:
From the failure we got, we can assume that it already happened.

Reply:
Also since we have turned off rollups for a while there are many entries that need to be rolled up.

Reply:
From the 2 databases I looked at that had this issue, the data was rolled up, but not deleted.

I verified by running a query and the amounts in the BW usage table, matched the rollup values for that hour.

If we had code that did an insert/on conflict update, the rollup would have been double the actual amount used, in this situation.

Maybe a solution would be to put a date range on the delete so we don't delete the data....It just gets left behind.  Then we can observe the frequency of the phenomenon. However, we could still be reporting bandwidth incorrectly.

Reply:
Why wasn't it deleted?

Reply:
Unfortunately I don't have enough info to answer that.  How about we add the date range to the delete so we don't delete the data that causes the constraint issue?

Then we can use that data to help us debug the issue and frequency of the problem.

Reply:
Let's try to recreate the situation before trying to fix it.

I just started wondering, what would happen if we set the rollup interval to every 5s.

Reply:
I can confirm that the database I saw had the same amounts in bandwidth_usage as in bandwidth_usage_rollup for the duplicate keys. This suggests the process got interrupted between the insert and the delete statement. You could consider wrapping the two into a transaction to prevent this from happening.

Reply:
Yes, both are possible causes, and both need fixes.

Reply:
An alternative:
```
	tx, err := db.Begin()
	if err != nil {
		return ErrInfo.Wrap(err)
	}

	defer func() {
		if err == nil {
			err = tx.Commit()
		} else {
			err = errs.Combine(err, tx.Rollback())
		}
	}()

	result, err := tx.Exec(`
		INSERT INTO bandwidth_usage_rollups (interval_start, satellite_id,  action, amount)
		SELECT datetime(strftime('%Y-%m-%dT%H:00:00', created_at)) created_hr, satellite_id, action, SUM(amount)
			FROM bandwidth_usage
		WHERE datetime(created_at) < datetime(?)
		GROUP BY created_hr, satellite_id, action
		ON CONFLICT(interval_start, satellite_id,  action)
		DO UPDATE SET amount = bandwidth_usage_rollups.amount + excluded.amount;

		DELETE FROM bandwidth_usage WHERE datetime(created_at) < datetime(?);
	`, hour, hour)
```
This uses a transaction and also allows for any straggling data after the rollup.

Reply:
You can include the BEGIN and COMMIT in the exec query.

Reply:
updated. leaving tx code as is, per our conversation